tests/integration/targets/vmware_ops_esxi_maintenance_mode_test/tasks/main.yml

```
@@ -0,0 +1,106 @@
+
+    - name: Provision Virtual Esxi
+      ansible.builtin.import_role:
+        name: cloud.vmware_ops.provision_virtual_esxi
+
+    - name: Wait until ESXi VM is powered off after OS installation
+      community.vmware.vmware_guest_info:
+        hostname: "{{ provision_virtual_esxi_hostname }}"
+        username: "{{ provision_virtual_esxi_username }}"
+        password: "{{ provision_virtual_esxi_password }}"
+        port: "{{ provision_virtual_esxi_port }}"
+        datacenter: "{{ provision_virtual_esxi_datacenter }}"
+        validate_certs: "{{ provision_virtual_esxi_validate_certs }}"
+        name: "{{ provision_virtual_esxi_vms[0].name }}"
+      register: _esxi_power_status
+      until: _esxi_power_status.instance.hw_power_status == 'poweredOff'
+      retries: 30
+      delay: 20
+
+    - name: Power on ESXi VM if it is off
+      ansible.builtin.include_role:
+        name: cloud.vmware_ops.provision_vm
+      vars:
+        provision_vm_state: "poweredon"
+
+    - name: Wait until ESXi VM has an IP address
+      community.vmware.vmware_guest_info:
+        hostname: "{{ provision_virtual_esxi_hostname }}"
+        username: "{{ provision_virtual_esxi_username }}"
+        password: "{{ provision_virtual_esxi_password }}"
+        port: "{{ provision_virtual_esxi_port }}"
+        datacenter: "{{ provision_virtual_esxi_datacenter }}"
+        validate_certs: "{{ provision_virtual_esxi_validate_certs }}"
+        name: "{{ provision_virtual_esxi_vms[0].name }}"
+      register: _esxi_host_check
+      until: _esxi_host_check.instance.ipv4 is defined and _esxi_host_check.instance.ipv4 is not none
+      retries: 30
+      delay: 10
+
+    - name: Check if ESXi is up
+      wait_for:
+        host: "{{ _esxi_host_check.instance.ipv4 }}"
+        port: 22
+        delay: 10
+        timeout: 300
+        state: started
+
+    - name: Add ESXI Host
+      ansible.builtin.import_role:
+        name: cloud.vmware_ops.vcenter_host_connection
+      vars:
+        vcenter_host_connection_state: present
+        vcenter_host_connection_esxi_hostname: "{{ _esxi_host_check.instance.ipv4 }}"
+
+    - name: Enable Maintenance Mode
+      ansible.builtin.import_role:
+        name: cloud.vmware_ops.esxi_maintenance_mode
+      vars:
+        esxi_maintenance_mode_esxi_hostname: "{{ _esxi_host_check.instance.ipv4 }}"
+        esxi_maintenance_mode_enable: True
+
+    - name: Disable Maintenance Mode
+      ansible.builtin.import_role:
+        name: cloud.vmware_ops.esxi_maintenance_mode
+      vars:
+        esxi_maintenance_mode_esxi_hostname: "{{ _esxi_host_check.instance.ipv4 }}"
+        esxi_maintenance_mode_enable: False
+
+  always:
+    - name: Remove ESXI Host
+      ansible.builtin.import_role:
+        name: cloud.vmware_ops.vcenter_host_connection
+      vars:
+        vcenter_host_connection_state: absent
+        vcenter_host_connection_esxi_hostname: "{{ _esxi_host_check.instance.ipv4 }}"
```

If a test failed before ip assigned or failed to get ip, all cleanup will fail because of undefined variable

```suggestion
        vcenter_host_connection_esxi_hostname: "{{ _esxi_host_check.instance.ipv4  | default("") }}"
```

---

fixed